### PR TITLE
wallet: call SyncWithValidationInterfaceQueue after disconnecting chain notifications

### LINKED ISSUE #34354: Windows CI failure: restore wallet test case related to pruned blockchain in `wallet_assumeutxo.py` is failing.
ISTM that the [Windows, ucrt, test cross-built CI failure](https://github.com/bitcoin/bitcoin/actions/runs/21172425203/job/60894641918?pr=33014) is unrelated, a restore wallet test case related to pruned blockchain in `wallet_assumeutxo.py` is failing.

<details open>
<summary>From the logs:</summary>

```zsh
2026-01-20T13:41:07.2385503Z 2026-01-20T13:41:04.152353Z TestFramework (INFO): Ensuring wallet can't be restored from a backup that was created before the pruneheight (pruned node)
2026-01-20T13:41:07.2386535Z 
2026-01-20T13:41:07.2386839Z 2026-01-20T13:41:06.479889Z TestFramework (ERROR): Unexpected exception
2026-01-20T13:41:07.2387306Z 
2026-01-20T13:41:07.2387503Z Traceback (most recent call last):
2026-01-20T13:41:07.2387815Z 
2026-01-20T13:41:07.2388269Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\util.py", line 166, in try_rpc
2026-01-20T13:41:07.2388901Z 
2026-01-20T13:41:07.2389049Z     fun(*args, **kwds)
2026-01-20T13:41:07.2389318Z 
2026-01-20T13:41:07.2389432Z     ~~~^^^^^^^^^^^^^^^
2026-01-20T13:41:07.2389660Z 
2026-01-20T13:41:07.2390180Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\coverage.py", line 50, in __call__
2026-01-20T13:41:07.2390808Z 
2026-01-20T13:41:07.2391156Z     return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
2026-01-20T13:41:07.2391651Z 
2026-01-20T13:41:07.2392093Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\authproxy.py", line 156, in __call__
2026-01-20T13:41:07.2392771Z 
2026-01-20T13:41:07.2392996Z     raise JSONRPCException(response['error'], status)
2026-01-20T13:41:07.2393424Z 
2026-01-20T13:41:07.2395532Z test_framework.authproxy.JSONRPCException: filesystem error: cannot remove: The process cannot access the file because it is being used by another process [D:\a\_temp\test_runner_₿_🏃_20260120_132914\wallet_assumeutxo_56\node3\regtest\w2\wallet.dat] (-1)
2026-01-20T13:41:07.2397271Z 
2026-01-20T13:41:07.2397277Z 
2026-01-20T13:41:07.2397281Z 
2026-01-20T13:41:07.2397613Z During handling of the above exception, another exception occurred:
2026-01-20T13:41:07.2398102Z 
2026-01-20T13:41:07.2398107Z 
2026-01-20T13:41:07.2398112Z 
2026-01-20T13:41:07.2398334Z Traceback (most recent call last):
2026-01-20T13:41:07.2398622Z 
2026-01-20T13:41:07.2399135Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\test_framework.py", line 142, in main
2026-01-20T13:41:07.2399796Z 
2026-01-20T13:41:07.2399922Z     self.run_test()
2026-01-20T13:41:07.2400161Z 
2026-01-20T13:41:07.2400286Z     ~~~~~~~~~~~~~^^
2026-01-20T13:41:07.2400474Z 
2026-01-20T13:41:07.2400948Z   File "D:\a\bitcoin\bitcoin/test/functional/wallet_assumeutxo.py", line 267, in run_test
2026-01-20T13:41:07.2401587Z 
2026-01-20T13:41:07.2401827Z     self.test_restore_wallet_pruneheight(n3)
2026-01-20T13:41:07.2402338Z 
2026-01-20T13:41:07.2402476Z     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
2026-01-20T13:41:07.2402806Z 
2026-01-20T13:41:07.2403384Z   File "D:\a\bitcoin\bitcoin/test/functional/wallet_assumeutxo.py", line 99, in test_restore_wallet_pruneheight
2026-01-20T13:41:07.2404151Z 
2026-01-20T13:41:07.2404691Z     assert_raises_rpc_error(-4, error_message, n3.restorewallet, "w2", "backup_w2.dat")
2026-01-20T13:41:07.2405294Z 
2026-01-20T13:41:07.2405553Z     ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-20T13:41:07.2405970Z 
2026-01-20T13:41:07.2406496Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\util.py", line 157, in assert_raises_rpc_error
2026-01-20T13:41:07.2407238Z 
2026-01-20T13:41:07.2407544Z     assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
2026-01-20T13:41:07.2408097Z 
2026-01-20T13:41:07.2408261Z            ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-01-20T13:41:07.2408598Z 
2026-01-20T13:41:07.2409048Z   File "D:\a\bitcoin\bitcoin\test\functional\test_framework\util.py", line 170, in try_rpc
2026-01-20T13:41:07.2409645Z 
2026-01-20T13:41:07.2410041Z     raise AssertionError("Unexpected JSONRPC error code %i" % e.error["code"])
2026-01-20T13:41:07.2410583Z 
2026-01-20T13:41:07.2410797Z AssertionError: Unexpected JSONRPC error code -1
2026-01-20T13:41:07.2411215Z 
2026-01-20T13:41:07.2411925Z 2026-01-20T13:41:06.484138Z TestFramework (INFO): Not stopping nodes as test failed. The dangling processes will be cleaned up later.
2026-01-20T13:41:07.2412795Z 
2026-01-20T13:41:07.2413676Z 2026-01-20T13:41:06.484270Z TestFramework (WARNING): Not cleaning up dir D:\a\_temp\test_runner_₿_🏃_20260120_132914\wallet_assumeutxo_56
2026-01-20T13:41:07.2414525Z 
2026-01-20T13:41:07.2415646Z 2026-01-20T13:41:06.484375Z TestFramework (ERROR): Test failed. Test logging available at D:\a\_temp\test_runner_₿_🏃_20260120_132914\wallet_assumeutxo_56/test_framework.log
2026-01-20T13:41:07.2416737Z 

2026-01-20T13:42:33.3464929Z wallet_assumeutxo.py                                                             | ✖ Failed  | 8 s
2026-01-20T13:42:33.3465209Z 
2026-01-20T13:42:33.3465460Z ALL                                                                              | ✖ Failed  | 2893 s (accumulated) 
2026-01-20T13:42:33.3465784Z Runtime: 796 s

```
</details>

_Originally posted by @rkrux in https://github.com/bitcoin/bitcoin/issues/33014#issuecomment-3773073528_

### PR DESCRIPTION
**Pull request by achow101**
When the wallet disconnects chain notifications, it is expecting no further notifications to execute, but this is not the case. This results in test failures such as in #34354. Instead of disconnecting the notifications and continuing shutdown, we should wait for the validation interface queue to be drained before the rest of wallet shutdown. This is achieved by adding an `interfaces::Chain::waitForNotifications()` function which calls `SyncWithValidationInterfaceQueue()`.

Fixes #34354

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -1072,6 +1072,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     //! Find the private key for the given key id from the wallet's descriptors, if available
     //! Returns nullopt when no descriptor has the key or if the wallet is locked.
     std::optional<CKey> GetKey(const CKeyID& keyid) const;
+
+    //! Disconnect chain notifications and wait for validation interfrace queue to drain
```
Comment (stickies-v) [inline src/wallet/wallet.h]: typo nit
```suggestion
    //! Disconnect chain notifications and wait for validation interface queue to drain
```

File: src/interfaces/handler.h [hunk 2]
```diff
@@ -26,6 +26,9 @@ class Handler
 
     //! Disconnect the handler.
     virtual void disconnect() = 0;
+
+    //! Block until all notifications have been handled
+    virtual void sync() = 0;
```
Comment (stickies-v) [inline src/interfaces/handler.h]: Adding a virtual `sync` method that's no-op for all but one of its implementations doesn't seem ideal. Would adding a `Chain::waitForNotifications` method be a workable alternative? E.g.:

<details>
<summary>git diff on 563ec53d54</summary>

```diff
diff --git a/src/common/interfaces.cpp b/src/common/interfaces.cpp
index 1c1a6f3203..ffd85e6131 100644
--- a/src/common/interfaces.cpp
+++ b/src/common/interfaces.cpp
@@ -17,7 +17,6 @@ public:
     explicit CleanupHandler(std::function<void()> cleanup) : m_cleanup(std::move(cleanup)) {}
     ~CleanupHandler() override { if (!m_cleanup) return; m_cleanup(); m_cleanup = nullptr; }
     void disconnect() override { if (!m_cleanup) return; m_cleanup(); m_cleanup = nullptr; }
-    void sync() override {}
     std::function<void()> m_cleanup;
 };
 
@@ -27,7 +26,6 @@ public:
     explicit SignalHandler(boost::signals2::connection connection) : m_connection(std::move(connection)) {}
 
     void disconnect() override { m_connection.disconnect(); }
-    void sync() override {}
 
     boost::signals2::scoped_connection m_connection;
 };
diff --git a/src/interfaces/chain.h b/src/interfaces/chain.h
index e6847b9b16..a0f288aee6 100644
--- a/src/interfaces/chain.h
+++ b/src/interfaces/chain.h
@@ -332,6 +332,9 @@ public:
     //! chain tip.
     virtual void waitForNotificationsIfTipChanged(const uint256& old_tip) = 0;
 
+    //! Wait for all pending notifications to be processed.
+    virtual void waitForNotifications() = 0;
+
     //! Register handler for RPC. Command is not copied, so reference
     //! needs to remain valid until Handler is disconnected.
     virtual std::unique_ptr<Handler> handleRpc(const CRPCCommand& command) = 0;
diff --git a/src/interfaces/handler.h b/src/interfaces/handler.h
index 6b8484f80f..b3aebf7e14 100644
--- a/src/interfaces/handler.h
+++ b/src/interfaces/handler.h
@@ -26,9 +26,6 @@ public:
 
     //! Disconnect the handler.
     virtual void disconnect() = 0;
-
-    //! Block until all notifications have been handled
-    virtual void sync() = 0;
 };
 
 //! Return handler wrapping a boost signal connection.
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index 3b2a726d10..2cc30ff519 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -499,10 +499,6 @@ public:
             m_proxy.reset();
         }
     }
-    void sync() override
-    {
-        m_signals.SyncWithValidationInterfaceQueue();
-    }
     ValidationSignals& m_signals;
     std::shared_ptr<NotificationsProxy> m_proxy;
 };
@@ -539,8 +535,6 @@ public:
             ::tableRPC.removeCommand(m_command.name, &m_command);
         }
     }
-    void sync() final {}
-
     ~RpcHandlerImpl() override { disconnect(); }
 
     CRPCCommand m_command;
@@ -790,6 +784,10 @@ public:
         if (!old_tip.IsNull() && old_tip == WITH_LOCK(::cs_main, return chainman().ActiveChain().Tip()->GetBlockHash())) return;
         validation_signals().SyncWithValidationInterfaceQueue();
     }
+    void waitForNotifications() override
+    {
+        validation_signals().SyncWithValidationInterfaceQueue();
+    }
     std::unique_ptr<Handler> handleRpc(const CRPCCommand& command) override
     {
         return std::make_unique<RpcHandlerImpl>(command);
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 9e918599d5..1e7ad3fc6b 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4582,7 +4582,7 @@ void CWallet::DisconnectChainNotifications()
 {
     if (m_chain_notifications_handler) {
         m_chain_notifications_handler->disconnect();
-        m_chain_notifications_handler->sync();
+        chain().waitForNotifications();
         m_chain_notifications_handler.reset();
     }
 }
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index b90668e64c..e6823c2441 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -1073,7 +1073,7 @@ public:
     //! Returns nullopt when no descriptor has the key or if the wallet is locked.
     std::optional<CKey> GetKey(const CKeyID& keyid) const;
 
-    //! Disconnect chain notifications and wait for validation interfrace queue to drain
+    //! Disconnect chain notifications and wait for validation interface queue to drain
     void DisconnectChainNotifications();
 };
 

```
</details>

**Review by stickies-v — COMMENTED** [review]
Concept ACK

File: src/wallet/wallet.h (hunk 1 above)
Comment (achow101) [reply src/wallet/wallet.h]: Done

File: src/interfaces/handler.h (hunk 2 above)
Comment (achow101) [reply src/interfaces/handler.h]: Taken the suggestion

**Review by furszy — COMMENTED** [review]
The current approach has a downside during IBD, any wallet unload during IBD could stall for hours because notification will keep coming. This wouldn't be much a deal for an RPC thread but for the GUI, this would totally freeze the app.

What we can do instead, is to unregister the wallet from the validation interface and submit a blocking task to the interface, so that we only wait up until the pending notifications gets cleared, and nothing else. Then we continue shutdown etc.

 I implemented it here https://github.com/furszy/bitcoin-core/commit/bf1abcd6d50374a461dad305f0243d3a613ce158. All yours (haven't tested it).

If we want more concurrency in the future, instead of waiting for the queue to drain, we can submit the remaining unload steps into the validation interface so they are executed only after the pending tasks get processed. I didn't do it because it would require further changes that don't think they are acceptable for the release freeze period.

**Comment by achow101** [comment]
> submit a blocking task to the interface, so that we only wait up until the pending notifications gets cleared, and nothing else. Then we continue shutdown etc.

Isn't that what `SyncWithValidationInterfaceQueue` does? It also calls `CallFunctionInValidationInterfaceQueue`, but with ` std::promise`, and then it `wait()`s on that future.

**Comment by furszy** [comment]
> Isn't that what `SyncWithValidationInterfaceQueue` does? It also calls `CallFunctionInValidationInterfaceQueue`, but with ` std::promise`, and then it `wait()`s on that future.

ha yeah.. nvm. Bad brain, dunno why I was completely convinced it was waiting until the queue was empty instead 🤷🏼‍♂️.

File: src/interfaces/chain.h [hunk 3]
```diff
@@ -332,6 +332,9 @@ class Chain
     //! chain tip.
     virtual void waitForNotificationsIfTipChanged(const uint256& old_tip) = 0;
 
+    //! Wait for all pending notifications to be processed
```
Comment (furszy) [inline src/interfaces/chain.h]: nano nit:

```suggestion
    //! Wait for all pending notifications up to this point to be processed
```

**Review by furszy — COMMENTED** [review]
ACK c010fe8a2f9ec0dcee09a872da8926e0ee5f91a4

File: src/interfaces/chain.h (hunk 3 above)
Comment (stickies-v) [reply src/interfaces/chain.h]: Or with even less ambiguity: "Wait for notifications that were pending before this call to be processed."

I don't think this is a nano nit, important to be precise when describing interfaces. Also would prefer improving the `handleNotifications` docstring to highlight that there may be pending notifications when the handler is destroyed, referencing `waitForNotifications`. This might prevent similar issues in the future. For example:

```cpp
    //! Register handler for notifications. Some notifications are asynchronous
    //! and may still execute after the handler is disconnected. Use
    //! waitForNotifications() after the handler is disconnected to ensure all
    //! pending notifications have been processed.
    virtual std::unique_ptr<Handler> handleNotifications(std::shared_ptr<Notifications> notifications) = 0;
```

note: if no `Notifications` are registered, there can be no pending notifications, yet this will still block, so the docstring is still not entirely correct. But it's consistent with pre-existing `waitForNotificationsIfTipChanged`, and I'm not sure there's currenly a better approach without significant scope creep.

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -1072,6 +1072,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     //! Find the private key for the given key id from the wallet's descriptors, if available
     //! Returns nullopt when no descriptor has the key or if the wallet is locked.
     std::optional<CKey> GetKey(const CKeyID& keyid) const;
+
+    //! Disconnect chain notifications and wait for validation interface queue to drain
```
Comment (stickies-v) [inline src/wallet/wallet.h]: nit: better adhere to interface boundaries/terminology
```suggestion
    //! Disconnect chain notifications and wait for all notifications to be processed.
```

Comment (stickies-v) [reply src/wallet/wallet.h]: An alternative approach would be to always wait for notifications to be processed in `disconnect()`. Is my understanding that the reason we don't is because during shutdown, the scheduler [may already have been stopped](https://github.com/bitcoin/bitcoin/blob/b6b8f8ac555f155a44ee3780b144e0960ae73483/src/init.cpp#L326)? If so, I think this would be helpful to document `DisconnectChainNotifications` better so this doesn't get refactored out in the future, it's quite subtle I think.

**Review by stickies-v — COMMENTED** [review]
crACK c010fe8a2f9ec0dcee09a872da8926e0ee5f91a4

PR description needs to be updated after the changed approach.

Waiting for the chain notifications to be processed makes the wallet shutdown more correct, but also slower. In most cases this seems harmless, but I did not test if there are edge cases (e.g. IBD, multiple wallets, a somehow clogged up/slow validation queue) where this can be problematic.

File: src/interfaces/chain.h (hunk 3 above)
Comment (furszy) [reply src/interfaces/chain.h]: @stickies-v, one option could be to move the listener unregistration into the interface so it mirrors the registration process (`handleNotifications`). That way we keep symmetry between the two and can enforce the requirements internally. See how I did it here https://github.com/furszy/bitcoin-core/commit/bf1abcd6d50374a461dad305f0243d3a613ce158 (note: the latch part can be replaced for a SyncWithQueue if not empty calls).

Comment (achow101) [reply src/interfaces/chain.h]: Updated the comment as suggested

I think it would be better if `disconnect()` did the waiting, but that would be enlarging the scope of this PR to be more than I think is appropriate for a feature freeze fix. Since it would affect more than just the wallet, there may be other side effects that have to be considered as well.

File: src/wallet/wallet.h (hunk 4 above)
Comment (achow101) [reply src/wallet/wallet.h]: Done as suggested.

**Review by furszy — COMMENTED** [review]
ACK 98e8af4bb991fd8edeb15c0fb8afa66bff6b5cac

File: src/wallet/wallet.h (hunk 4 above)
Comment (stickies-v) [reply src/wallet/wallet.h]: > If so, I think this would be helpful to document `DisconnectChainNotifications` better so this doesn't get refactored out in the future, it's quite subtle I think.

It's minor, but I think this comment has not yet been addressed. If my understanding is correct, a suggestion could be:

<details>
<summary>git diff on 98e8af4bb9</summary>

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 1e7ad3fc6b..4fa6203d6d 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4580,6 +4580,9 @@ std::optional<WalletTXO> CWallet::GetTXO(const COutPoint& outpoint) const
 
 void CWallet::DisconnectChainNotifications()
 {
+    // waitForNotifications() is called separately from disconnect() rather than integrated into it,
+    // because Shutdown() may tear down the scheduler before notifications are disconnected, causing
+    // the disconnection to block.
     if (m_chain_notifications_handler) {
         m_chain_notifications_handler->disconnect();
         chain().waitForNotifications();

```
</details>

**Comment by stickies-v** [comment]
utACK 98e8af4bb991fd8edeb15c0fb8afa66bff6b5cac

File: src/interfaces/chain.h (hunk 3 above)
Comment (rkrux) [reply src/interfaces/chain.h]: > Some notifications are asynchronous

Shouldn't majority of the notifications in the codebase be asynchronous? Considering that the queueing mechanism is used usually used for async stuff (not specific to core).

And I assume synchronous notifications are those that are called like how `SyncWithValidationInterfaceQueue` calls? 
https://github.com/bitcoin/bitcoin/blob/9cad97f6cdf1bd660732cd10e844a6a7e0771ea0/src/validationinterface.cpp#L146-L155

**Review by rkrux — APPROVED** [review]
crACK 98e8af4bb991fd8edeb15c0fb8afa66bff6b5cac

This makes the wallet unloading / shutdown operation more robust albeit slower (specially in case of multiple wallets loaded), which I don't mind. I'd prefer consistent & slow(er) operations over brittle & fast(er) ones. 

I too feel that the waiting portion can be encapsulated within the disconnect function, but can be addressed later.

**Review by sedited — APPROVED** [review]
ACK 98e8af4bb991fd8edeb15c0fb8afa66bff6b5cac
